// ngx-actioncable/esm2020/ngx-action-cable/ngx-action-cable.service.mjs
import { Injectable, Optional } from '@angular/core';
import { NgxActionCableBroadcaster } from './ngx-action-cable-broadcaster';
import * as ActionCable from 'actioncable';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-action-cable-configuration";
export class NgxActionCableService {
    constructor(configuration) {
        this.configuration = configuration;
        this.cables = new Map();
        this.channels = {};
        // TODO: remove this workaround -> createWebSocketURL is undefined exception 
        // in action_cable.js
        let w = window;
        w.createWebSocketURL = ActionCable.createWebSocketURL;
        configuration.urls.forEach((url, key) => {
            let cable = this.connect(key, url);
            this.cables.set(key, cable);
        });
    }
    subscribe(key, channel, params = {}) {
        let channelName = this.getChannelName(channel, params);
        let subscriptionParams = Object.assign({ channel: channel }, params);
        let broadcaster = new NgxActionCableBroadcaster();
        let cable = this.getCabel(key);
        let subscription = cable.subscriptions.create(subscriptionParams, {
            received: (data) => {
                broadcaster.broadcast(data.action, data);
            }
        });
        this.channels[channelName] = {
            subscription: subscription,
            broadcaster: broadcaster
        };
        return broadcaster;
    }
    unsubscribe(key, channel, params = {}) {
        let channelName = this.getChannelName(channel, params);
        let cable = this.getCabel(key);
        if (!this.channels[channelName]) {
            console.info(`No Subscription for Channel ${channelName} found!`);
        }
        else {
            let subscription = this.channels[channelName].subscription;
            cable.subscriptions.remove(subscription);
        }
    }
    perform(channel, params = {}, action, data) {
        let channelName = this.getChannelName(channel, params);
        this.channels[channelName].subscription.perform(action, data);
    }
    connect(key, url) {
        let cable = ActionCable.createConsumer(url);
        cable.connect();
        return cable;
    }
    disconnect(key) {
        let cable = this.getCabel(key);
        cable.disconnect();
    }
    getChannelName(channel, params = {}) {
        let channelName = (typeof (channel) === 'object') ? channel['channel'] : channel;
        channelName += `_${JSON.stringify(params)}`; // also add params to unique channel name
        return channelName;
    }
    getCabel(key) {
        let cable = this.cables.get(key);
        if (!cable) {
            throw Error(`No cable instance for key ${key} found!`);
        }
        return cable;
    }
}
NgxActionCableService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.4", ngImport: i0, type: NgxActionCableService, deps: [{ token: i1.NgxActionCableConfiguration, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
NgxActionCableService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.4", ngImport: i0, type: NgxActionCableService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.4", ngImport: i0, type: NgxActionCableService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.NgxActionCableConfiguration, decorators: [{
                    type: Optional
                }] }]; } });
