// ngx-actioncable/esm2020/ngx-action-cable/ngx-action-cable.module.mjs
import { NgxActionCableConfiguration } from './ngx-action-cable-configuration';
import { NgxActionCableService } from './ngx-action-cable.service';
import { NgModule } from '@angular/core';
import * as i0 from "@angular/core";
export class NgxActionCableModule {
    static forConfig(configurationFactory) {
        return {
            ngModule: NgxActionCableModule,
            providers: [{ provide: NgxActionCableConfiguration, useFactory: configurationFactory }]
        };
    }
}
NgxActionCableModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.4", ngImport: i0, type: NgxActionCableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxActionCableModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.4", ngImport: i0, type: NgxActionCableModule });
NgxActionCableModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.4", ngImport: i0, type: NgxActionCableModule, providers: [
        NgxActionCableService
    ], imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.4", ngImport: i0, type: NgxActionCableModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    providers: [
                        NgxActionCableService
                    ]
                }]
        }] });
